### ISSUE TITLE
Panic "not supported" during parsing

### ISSUE THREAD

**Issue by aclemmensen** [mischov/meeseeks#27]
I found another one; not sure how helpful it is as the message does seem to indicate it's intentional:

```
thread '<unnamed>' panicked at 'not supported', src/flat_dom.rs:320:8
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:380
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:396
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:611
   5: std::panicking::begin_panic_new
   6: <html5ever::tree_builder::TreeBuilder<Handle, Sink> as html5ever::tree_builder::rules::TreeBuilderStep<Handle>>::step
   7: <html5ever::tree_builder::TreeBuilder<Handle, Sink> as html5ever::tokenizer::interface::TokenSink>::process_token
   8: <html5ever::tokenizer::Tokenizer<Sink>>::process_token
   9: <html5ever::driver::Parser<Sink> as tendril::stream::TendrilSink<tendril::fmt::UTF8>>::finish
  10: tendril::stream::TendrilSink::one
  11: std::panicking::try::do_call
  12: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  13: <F as scoped_pool::Task>::run
  14: scoped_pool::Pool::run_thread
```

I don't have a repro case for this one yet but I could probably dig one up if you're interested.

**Comment by mischov** [comment]
Looks like it's script related (`mark_script_already_started`)- that isn't something I'm handling currently. I can look at feasibility.

I know you mentioned that the previous error brought the entire app down- are you checking for an `{:error, cause}` tuple after you either parse or select, but before you try to extract from the parsed document?

**Comment by aclemmensen** [comment]
Yeah, I was suspecting as much. I made a little test case with nested script tags, but that didn't cause this panic to show up.

Regarding taking the entire app down, I found out I was a bit quick in my judgment there and misread some error log output. It turns out that the app does indeed survive the panic. I experienced the same when looking for a repro for my previous report; it would print the panic on output but still continue parsing documents after the error. It seems this case is similar - at least I see the same error printed multiple times in the stderr output of my service, quite likely because I'm parsing many pages in a row from the same website containing the same odd HTML.

Regarding the return value from `Meeseeks.parse()`, I don't actually check for error tuples and I haven't encountered any errors relating to the type. I just feed the output directly into `Meeseeks.all` with a selector. I also don't need to extract anything, as my use-case is only to detect whether a page matches a set of selectors and then return the number of occurrences of each match.

I'll try to find a reliable repro for this panic and update this issue accordingly.

**Comment by mischov** [comment]
Okay. If you're using `Meeseeks.parse` before `Meeseeks.all` then you can match the return value of parse like:
```elixir
case Meeseeks.parse(html) do
  {:error, cause} -> Logger.error "Unable to parse HTML: #{cause}"
  document -> Meeseeks.all(document, ...)
end
```

**Comment by aclemmensen** [comment]
Ah, cool, I didn't notice that. I just checked the docs and the error tuple isn't in any of the examples but it is indeed included in the method's spec. I'll make sure to add the necessary checking. Thank you.

**Comment by mischov** [comment]
Yeah, I haven't included it in any examples- sounds like something I should probably remedy. `all`, `one` and `select` also pass through parse errors which is why I was wondering if you were trying to extract from an error and raising.

If you're commonly hitting things I've chosen to panic on as `unsupported`, maybe I need to consider treating them as no-ops and trying to figure out how to sensibly warn people? I have basically *never* hit one of them so it's not been on my radar.

**Comment by mischov** [comment]
I appreciate the feedback, by the way. It's very helpful.

**Comment by aclemmensen** [comment]
I'd support having it in the examples. Showing the best way to do it up front is generally the nicest way to learn how to use a library, I'd say.

My use-case is basically:

```elixir
Meeseeks.parse(content, :html) |> Meeseeks.all(css("some-selector")) |> Enum.count
```

I'd also support having unsupported cases as no-ops. At least to the extent that the operation is meaningless. Having panics or assertions is good when testing and verifying compliance with a spec, but they do have a tendency to throw a wrench in the works in a running application. Perhaps there's a way to make the panic vs. no-op configurable? A sort of "strict mode".

**Comment by mischov** [comment]
As an aside, (I don't suggest doing this because the savings aren't going to be that large unless you're working on a truly massive scale) you could absolutely write a custom accumulator that just increments on each match and save yourself the construction of results and the `Enum.count` at the end.

**Comment by aclemmensen** [comment]
It might actually be worth it. The project I'm working on is an attempt at moving away from an unmaintained C# HTML parser library to something more performant. So performance was definitely part of the consideration in looking at Meeseeks/html5ever. And the way the system works I have to parse millions of pages per day and check them against a fairly stable set of selectors.

**Comment by mischov** [comment]
Configuring between panic and no-op is an interesting idea. I was currently panicking because I didn't want Meeseeks returning an unexpected result because it silently ignored some part of the tree-building process, but as you say... a wrench in a running application.

**Comment by mischov** [comment]
Don't know if you've looked at accumulators but [they're not particularly complicated](https://github.com/mischov/meeseeks/blob/master/lib/meeseeks/accumulator/all.ex), and you'd just use [`select`](https://github.com/mischov/meeseeks/blob/master/lib/meeseeks.ex#L243-L262) instead of `all`.

**Comment by mischov** [comment]
Problem should be fixed as of `v0.7.6`, as should another possible panic before you get to that one!

I don't actually have a test for `mark_script_already_started` (which is now removed, so should be impossible to hit anyway), so if you somehow hit that error again please re-open the issue.
